cmd/influxd/export/export.go

```diff
@@ -0,0 +1,367 @@
+	}
+	cmd.path = fs.Arg(0)
+
+	err = os.MkdirAll(cmd.path, 0700)
+
+	return
+}
+
+// exportShard will write a tar archive of the passed in shard with any TSM files that have been
+// created since the time passed in
+func (cmd *Command) exportShard(retentionPolicy string, shardID string, tsStart, tsEnd time.Time) error {
+	id, err := strconv.ParseUint(shardID, 10, 64)
+	if err != nil {
+		return err
+	}
+
+	shardArchivePath, err := cmd.nextPath(filepath.Join(cmd.path, fmt.Sprintf(BackupFilePattern, cmd.database, retentionPolicy, id)))
+	if err != nil {
+		return err
+	}
+
+	cmd.StdoutLogger.Printf("backing up db=%v rp=%v shard=%v to %s start %s end  %s",
+		cmd.database, retentionPolicy, shardID, shardArchivePath, tsStart, tsEnd)
+
+	req := &snapshotter.Request{
+		Type:            snapshotter.RequestShardExport,
+		Database:        cmd.database,
+		RetentionPolicy: retentionPolicy,
+		ShardID:         id,
+		ExportStart:     tsStart,
+		ExportEnd:       tsEnd,
+	}
+
+	// TODO: verify shard backup data
+	return cmd.downloadAndVerify(req, shardArchivePath, nil)
+}
+
+// exportDatabase will request the database information from the server and then backup the metastore and
+// every shard in every retention policy in the database. Each shard will be written to a separate tar.
+func (cmd *Command) exportDatabase(tsStart, tsEnd time.Time) error {
+	cmd.StdoutLogger.Printf("backing up db=%s start %s end %s", cmd.database, tsStart, tsEnd)
+
+	req := &snapshotter.Request{
+		Type:     snapshotter.RequestDatabaseInfo,
+		Database: cmd.database,
+	}
+	cmd.StdoutLogger.Println("made it to before the export response paths")
+	response, err := cmd.requestInfo(req)
+	if err != nil {
+		return err
+	}
+	cmd.StdoutLogger.Println("made it to the export response paths")
+	return cmd.exportResponsePaths(response, tsStart, tsEnd)
+}
+
+// exportRetentionPolicy will request the retention policy information from the server and then backup
+// the metastore and every shard in the retention policy. Each shard will be written to a separate tar.
+func (cmd *Command) exportRetentionPolicy(retentionPolicy string, tsStart, tsEnd time.Time) error {
+	cmd.StdoutLogger.Printf("backing up rp=%s start %s end %s", retentionPolicy, tsStart, tsEnd)
+
+	req := &snapshotter.Request{
+		Type:            snapshotter.RequestRetentionPolicyInfo,
+		Database:        cmd.database,
+		RetentionPolicy: retentionPolicy,
+	}
+
+	response, err := cmd.requestInfo(req)
+	if err != nil {
+		return err
+	}
+
+	return cmd.exportResponsePaths(response, tsStart, tsEnd)
+}
+
+// exportResponsePaths will backup the metastore and all shard paths in the response struct
+func (cmd *Command) exportResponsePaths(response *snapshotter.Response, tsStart, tsEnd time.Time) error {
+	if err := cmd.exportMetastore(); err != nil {
+		return err
+	}
+
+	// loop through the returned paths and back up each shard
+	for _, path := range response.Paths {
+		rp, id, err := retentionAndShardFromPath(path)
```

Comment:
I think this approach is very wasteful. Your'e requesting the database info, which is really just a bunch of relative shard paths, and then parsing those paths to get the rp name and shard ID. All of this while having a full binary metadata snapshot that can be decoded and contains that information for all databases.

I think the overall approach of this command should take that into account: grab the latest metadata snapshot, and filter shards based on that. You'll even have local access to the timestamps for each shard, and won't need to return "empty" shards from the server.

Reply:
I'm going to prioritize other work and make a ticket for this.  #8964